### DISCUSSION TITLE
NPM/PNPM support & preference for Romcal

### DISCUSSION THREAD

**Discussion by TobiTenno** [romcal/romcal#494] [Ideas]
@tukusejssirs and I have been running across some interesting side effects when using pnpm (as is @tukusejssirs' preference).

I'm of the opinion that with something like this that's cross-compiled to multiple platforms, I don't see the benefit to our lifecycle or maintenance to switching to pnpm on a regular/recommended basis.

The downsides of this, in my initial view (and we can have a whole thread for each of these):
1. using an alternative to `npm` can lead to unexpected script or performance behavior compared to those of our consumers
2. `pnpm` has shown to have odd behaviors (see discussions on https://github.com/romcal/romcal-examples/pull/2 and https://github.com/romcal/romcal/pull/479) that seem to break `npm` expectations
3. `pnpm` skips lifecycle steps by default
	- fully admit this is a preference, but ties back into (1)

**Comment by tukusejssirs** [comment]
In general, I think we should try to support NPM, PNPM and possibly Yarn in all romcal packages.

As for performance degradation (point 1): IMHO the main issue with the custom build script is that it needs to process each calendar + locale bundle individually which needs to be done _before_ loading the website. How about loading the website and the calendars would be loaded lazily? That is, first load the default calendar (which ATM is the GRC in English), then load the rest.

As for the PNPM requiring exact versioning of the `dev` romcal packages: it might be caused by the fact that we already have a `romcal@3.0.0` version released at NPM, albeit deprecated.

As for PNPM skipping steps: I believe you are talking about `pre*` and `post*` scripts. That is easily fixable by commiting `.npmrc` with `enable-pre-post-scripts=true`. Actually, I usually add the following into `.npmrc` (usually using `pnpm c --location project set "$config_name" "$config_value"`):

```ini
auto-install-peers=true
dedupe-peer-dependents=true
enable-pre-post-scripts=true
save-exact=true  # This is my preference, but we might want to remove it in romcal
strict-peer-dependencies=false
use-lockfile-v6=true
```

**Reply by TobiTenno** [reply]
supporting is different from preferring, in terms of what we use in the project itself.

supporting pnpm on breaking behaviors that it does (arguably wrong if it causes an issue) seem to be pnpm's issue, not romcal's

**Reply by emagnier** [reply]
Yes, indeed, we need to choose a package manager for Romcal, and if we stick with `npm`, I don't see any issue with that, as long as it meets our needs.
We also mention in several places about transitioning the structure of this project to a monorepo, so the selected package manager should be able to accommodate this requirement as well.
